spring-cloud-kubernetes-commons/src/main/java/org/springframework/cloud/kubernetes/commons/config/NormalizedSource.java

```diff
@@ -0,0 +1,80 @@
+/*
+ * Copyright 2013-2021 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.cloud.kubernetes.commons.config;
+
+/**
+ * Base class for Normalized Sources. It should contain all the "normalized" properties
+ * that users can specify, either explicitly or implicitly.
+ *
+ * @author wind57
+ */
+public sealed abstract class NormalizedSource permits NamedSecretNormalizedSource,LabeledSecretNormalizedSource,NamedConfigMapNormalizedSource {
+
+	private final String namespace;
+
+	private final String name;
+
+	private final boolean failFast;
+
+	private boolean frozen;
+
+	protected NormalizedSource(String name, String namespace, boolean failFast) {
+		this.name = name;
+		this.namespace = namespace;
+		this.failFast = failFast;
+	}
+
+	// this can return null, which is perfectly fine.
+	public final String namespace() {
+		if (frozen) {
+			throw new IllegalArgumentException("namespace must be queried elsewhere");
+		}
+		return this.namespace;
+	}
+
+	// this can return null (for a secret based on labels, for example)
+	public final String name() {
+		return this.name;
+	}
+
+	public final boolean failFast() {
+		return failFast;
+	}
+
+	/**
+	 * type of this normalized source. Callers are sensitive towards the actual type
+	 * specified.
+	 */
+	public abstract NormalizedSourceType type();
+
+	public abstract String target();
+
+	/**
+	 * calling this method will result in an Exceptions being thrown when calling some
+	 * getters. At the moment, this is done for the namespace only.
+	 */
+	public final void freeze() {
```

Comment:
Not sure I understand what this does....I must be missing something :)

Reply:
you're not. I did not plan for this, initially. I only added it after making the mistake a couple of times and then not understanding why tests are failing. 

So:

- users can specify a namespace, this will be translated in our codebase into a field in the `NormalizedSource`
- `NormalizedSource` is a field in `Fabric8ConfigContext`
- In `Fabric8SecretsPropertySourceLocator` we need to find out the "real" namespace (if `NormalizedSource` does not contain it), as such:

```
String namespace = getApplicationNamespace(client, normalizedSource.namespace(), normalizedSource.target(),
				provider);
		normalizedSource.freeze();
		Fabric8ConfigContext context = new Fabric8ConfigContext(client, normalizedSource, namespace, environment);
``` 

Now think about for the classes that get a `Fabric8ConfigContext` as input. They need the namespace to be able to make the actual requests. They have a choice:

- get it from `NormalizedSource::namespace`
- get it directly as a field in `Fabric8ConfigContext`

which one is correct? The second one is, because the first one might be missing entirely. 

In order for the implementations not to make this mistake (this is where I failed a couple of times), I "freeze" the NormalizedSource, since the namespace computation (the final one - the one that we will be using to actually read the configmap/secret) was done, so reading the namespace from it, is actually incorrect. 

Sorry for the long wording.

Reply:
Thanks for the explanation!

I think it might be better to encapsulate the logic behind where to get the namespace from in a method in the `ConfigContext`.

Reply:
I thought about that, and that is not that easy. It will again take a lot of words explaining why, though. If you are against this freezing (which I totally get), we can drop it and just enforce the fact that reading the namespace should be done from `Fabric8ConfigContext`, via unit tests. 

This freezing, in general, is "kind of" a protection. You can easily _not_ call `NormalizedSource::freeze` and of course can call `NormalizedSource::namespace` afterwards, so I am more than OK dropping this freezing.

Reply:
Yes lets drop it.

Reply:
will do! thx. I just can't build this currently because of [this issue](https://github.com/spring-cloud/spring-cloud-kubernetes/pull/985). Just FYI